### PR TITLE
vdk-kerberos-auth: add functional tests

### PR DESCRIPTION
**Pull request by tpalashki**
The newly introduced kerberos-auth plugin was missing tests.
This commit adds functional tests that run against a locally
started Kerberos server.

Signed-off-by: Tsvetomir Palashki <tpalashki@vmware.com>

### CODE REVIEW

File: projects/vdk-plugins/vdk-test-utils/src/vdk/plugin/test_utils/util_funcs.py [hunk 1]
```diff
@@ -135,6 +135,14 @@ def jobs_path_from_caller_directory(job_name: str):
     return get_test_job_path(caller_dir, job_name)
 
 
+def get_caller_directory():
+    import inspect
+
+    frame = inspect.stack()[1]
+    module = inspect.getmodule(frame[0])
+    return pathlib.Path(os.path.dirname(os.path.abspath(module.__file__)))
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-test-utils/src/vdk/plugin/test_utils/util_funcs.py:143]: this line could be reused(composed) in favour of copy-paste from another part of util_funcs.py

File: projects/vdk-plugins/.plugin-common.yml [hunk 2]
```diff
@@ -28,6 +28,16 @@
     reports:
       junit: tests.xml
 
+.build-plugin-dind:
```
[the remark this answers is not in the document]
Comment (Mepemotrarial) [reply projects/vdk-plugins/.plugin-common.yml:31]: I will create a PR later to change all other plugins.

File: projects/vdk-plugins/vdk-kerberos-auth/src/vdk/plugin/kerberos/krb5.conf [hunk 3]
```diff
@@ -1,15 +1,8 @@
...
 
 [realms]
-  PHONEHOME.VMWARE.COM = {
-      kdc = server01.vmware.com
-      kdc = server02.vmware.com
-      master_kdc = server.vmware.com
-      admin_server = server.vmware.com
+  EXAMPLE.COM = {
+    kdc = localhost
+    admin_server = localhost
```
[the remark this answers is not in the document]
Comment (Mepemotrarial) [reply projects/vdk-plugins/vdk-kerberos-auth/src/vdk/plugin/kerberos/krb5.conf:7]: This is the 'default' krb5.conf file to fall back to if no krb5.conf file is specified.

File: projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+def docker_compose_file(pytestconfig):
+    return os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), "docker-compose.yml"
+    )
+
+
+@pytest.fixture(scope="session")
+def kerberos_service(docker_ip, docker_services):
+    """Ensure that Kerberos server is up and responsive."""
+    time.sleep(10)
```
[the remark this answers is not in the document]
Comment (Mepemotrarial) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py:22]: I could not figure out how to check this particular server for responsiveness. I will dig a bit further.

File: projects/vdk-plugins/vdk-test-utils/src/vdk/plugin/test_utils/util_funcs.py (hunk 1 above)
Comment (Mepemotrarial) [reply projects/vdk-plugins/vdk-test-utils/src/vdk/plugin/test_utils/util_funcs.py:143]: The problem with reusing this in, say, jobs_path_from_caller_directory is that the stack trace would change. I will see if I can make it work both when called directly and indirectly, via another function.

File: projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    krb5_pass = "iamaserver"  # nosec
+    with mock.patch.dict(
+        os.environ,
+        {
+            "KRB5_CONFIG": str(caller_directory.joinpath("krb5.conf")),
+        },
+    ):
+        if os.path.isfile(keytab_filename):
+            os.remove(keytab_filename)
+        call(  # nosec
```
[the remark this answers is not in the document]
Comment (Mepemotrarial) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py:43]: Good point, I will see if I also need to check the return code.

File: projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                "kadmin",
+                "-p",
+                "admin/admin",
+                "-w",
+                krb5_pass,
+                "-q",
+                f"add_principal -randkey pa__view_{datajob_name}",
+            ]
+        )
+        call(  # nosec
```
[the remark this answers is not in the document]
Comment (Mepemotrarial) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py:54]: I will check.

File: projects/vdk-plugins/vdk-kerberos-auth/tests/test_kerberos.py [hunk 7]
```diff
@@ -17,4 +23,36 @@ def test_no_authentication(self):
...
+            {
+                "VDK_KRB_AUTH": "kinit",
+                "VDK_KRB5_CONF_FILENAME": krb5_conf_filename,
+            },
+        ):
+            result: Result = self.__runner.invoke(
+                ["run", jobs_path_from_caller_directory("test-job")]
+            )
+
+            cli_assert_equal(0, result)
```
[the remark this answers is not in the document]
Comment (Mepemotrarial) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/test_kerberos.py:41]: Can you elaborate on why this is a problem?

File: projects/vdk-plugins/.plugin-common.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tpalashki) [reply projects/vdk-plugins/.plugin-common.yml:31]: I will create a PR later to change all other plugins.

File: projects/vdk-plugins/vdk-kerberos-auth/src/vdk/plugin/kerberos/krb5.conf (hunk 3 above)
[the remark this answers is not in the document]
Comment (tpalashki) [reply projects/vdk-plugins/vdk-kerberos-auth/src/vdk/plugin/kerberos/krb5.conf:7]: This is the 'default' krb5.conf file to fall back to if no krb5.conf file is specified.

File: projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tpalashki) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py:22]: I could not figure out how to check this particular server for responsiveness. I will dig a bit further.

File: projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tpalashki) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py:43]: Good point, I will see if I also need to check the return code.

File: projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tpalashki) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/conftest.py:54]: I will check.

File: projects/vdk-plugins/vdk-kerberos-auth/tests/test_kerberos.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tpalashki) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/test_kerberos.py:41]: Can you elaborate on why this is a problem?

File: projects/vdk-plugins/vdk-test-utils/src/vdk/plugin/test_utils/util_funcs.py (hunk 1 above)
Comment (tpalashki) [reply projects/vdk-plugins/vdk-test-utils/src/vdk/plugin/test_utils/util_funcs.py:143]: The problem with reusing this in, say, `jobs_path_from_caller_directory` is that the stack trace would change. I will see if I can make it work both when called directly and indirectly, via another function.

**Review by mivanov1988 — APPROVED** [review]

File: projects/vdk-plugins/vdk-kerberos-auth/tests/test_kerberos.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ivakoleva) [reply projects/vdk-plugins/vdk-kerberos-auth/tests/test_kerberos.py:41]: The concept of testing any authentication, is confirming that auth fails when no valid credentials provided. Then, verifying having the correct credentials the authentication is successful.

Each `test_no_authentication()`,  `test_kinit_authentication()` and `test_minikerberos_authentication()` tests 
implementation currently confirm the data job runs with no error:
```
        result: Result = self.__runner.invoke(
             ["run", jobs_path_from_caller_directory("test-job")]
         )

         assert result.exit_code == 0
         cli_assert_equal(0, result)
```
The only difference is in some env variables; how are we sure the authentication is actually triggered when those env vars are set - and the test is false-positive due falling back to default `test_no_authentication()` use-case?

My suggestion is, for happy path -> to intercept authentication is actually triggered and successful, so it results in data job's successful execution;
Also, add a scenario with auth enabled yet authentication by keytab fails - then verify the data job does not run after - except results in a specific error expected;

**Review by ivakoleva — CHANGES_REQUESTED** [review]

**Comment by tpalashki** [comment]
I have added more test to cover additional use cases, such as failures due to various reasons (missing keytab or invalid credentials). I've also introduced additional output validations to each test.

**Comment by tpalashki** [comment]
> LGTM; Locally, after the conftest.py changes regarding service responsiveness, I cannot run the tests due to service timeouts. I guess you don't experience this problem locally, and remotely tests seem to be running just fine

Does the 90s timeout occur when you run locally?

**Comment by ivakoleva** [comment]
> > LGTM; Locally, after the conftest.py changes regarding service responsiveness, I cannot run the tests due to service timeouts. I guess you don't experience this problem locally, and remotely tests seem to be running just fine
> 
> Does the 90s timeout occur when you run locally?

Yup, it seems to honour the 90s timeout; issue seems to reproduce on a particular local environment only, therefore not a show-stopper, would suggest to merge the change;

Nevertheless, applying a stacktrace for visibility:
```

tests/conftest.py:68: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Services(_docker_compose=DockerComposeExecutor(_compose_files=['/Users/ikoleva/IdeaProjects/versatile-data-kit/projects/vdk-plugins/vdk-kerberos-auth/tests/docker-compose.yml'], _compose_project_name='pytest4518'), _services={})
check = <function _is_responsive at 0x10cfe3710>, timeout = 90.0, pause = 0.3
clock = <built-in function perf_counter>

    def wait_until_responsive(self, check, timeout, pause, clock=timeit.default_timer):
        """Wait until a service is responsive."""
    
        ref = clock()
        now = ref
        while (now - ref) < timeout:
            if check():
                return
            time.sleep(pause)
            now = clock()
    
>       raise Exception("Timeout reached while waiting on service!")
E       Exception: Timeout reached while waiting on service!
```

**Pull request merged**